src/thorin/pass/rw/clos2sjlj.cpp

```diff
@@ -0,0 +1,173 @@
+#include "thorin/pass/rw/clos2sjlj.h"
+#include "thorin/transform/clos_conv.h"
+
+namespace thorin {
+
+void Clos2SJLJ::get_exn_closures(const Def* def, DefSet& visited) {
+    if (def->level() != Sort::Term || def->isa_nom<Lam>() || visited.contains(def)) return;
+    visited.emplace(def);
+    if (auto c = isa_clos_lit(def)) {
+        auto lam = c.fnc_as_lam();
+        if (c.is_basicblock() && !ignore_.contains(lam)) {
+            def->world().DLOG("FOUND exn closure: {}", c.fnc_as_lam());
+            lam2tag_[c.fnc_as_lam()] = {lam2tag_.size() + 1, c.env()};
+        }
+        get_exn_closures(c.env(), visited);
+    } else {
+        for (auto op: def->ops())
+            get_exn_closures(op, visited);
+    }
+}
+
+void Clos2SJLJ::get_exn_closures() {
+    lam2tag_.clear();
+    if (!curr_nom()->is_set() || !curr_nom()->type()->is_cn()) return;
+    auto app = curr_nom()->body()->isa<App>();
+    if (!app) return;
+    if (auto p = app->callee()->isa<Extract>(); p && isa_clos_type(p->tuple()->type())) {
+        auto p2 = p->tuple()->isa<Extract>();
+        if (p2 && p2->tuple()->isa<Tuple>()) {
+            // branch: Check the closure environments, but be careful not to traverse
+            // the closures themselves
+            auto branches = p2->tuple()->ops();
+            for (auto b: branches) {
+                auto c = isa_clos_lit(b);
+                if (c) {
+                    ignore_.emplace(c.fnc_as_lam());
+                    world().DLOG("IGNORE {}", c.fnc_as_lam());
+                }
+            }
+        }
+    }
+    auto visited = DefSet();
+    get_exn_closures(app->arg(), visited);
+}
+
+static std::array<const Def*, 3> split(const Def* def) {
+    auto new_ops = DefArray(def->num_projs() - 2, nullptr);
+    auto& w = def->world();
+    const Def* mem, *env;
+    auto j = 0;
+    for (size_t i = 0; i < def->num_projs(); i++) {
+        auto op = def->proj(i);
+        if (op == w.type_mem() || op->type() == w.type_mem())
+            mem = op;
+        else if (i == CLOS_ENV_PARAM)
+            env = op;
+        else
+            new_ops[j++] = op;
+    }
+    assert(mem && env);
+    auto remaining = (def->level() == Sort::Term) ? w.tuple(new_ops) : w.sigma(new_ops);
+    return {mem, env, remaining};
+}
+
+static const Def* rebuild(const Def* mem, const Def* env, const Def* remaining) {
+    auto& w = mem->world();
+    auto env_seen = false;
+    auto new_ops = DefArray(remaining->num_projs() + 2, [&](auto i) {
+        if (i == 0) return mem;
+        if (i == 1) { env_seen = true; return env; }
+        return remaining->proj(env_seen ? i - 1 : i - 2);
+    });
+    return (remaining->level() == Sort::Term) ? w.tuple(new_ops) : w.sigma(new_ops);
+}
+
+Lam* Clos2SJLJ::get_throw(const Def* dom) {
+    auto& w = world();
+    auto [p, inserted] = dom2throw_.emplace(dom, nullptr);
+    auto& tlam = p->second;
```

Comment:
double-check that no rehash occurs inbetween. Otherwise, bad things will happen, if you modify the content pointed to by `tlam`.